From 5af11094e01bec3b4e0ea607d2ced76bd71bdf5f Mon Sep 17 00:00:00 2001
From: Dong Aisheng <aisheng.dong@nxp.com>
Date: Thu, 7 Sep 2017 14:15:11 +0800
Subject: [PATCH 3324/5242] MLK-17491-29 clk: imx: clk-pfdv2: fix the wrong
 pfd rate exported

commit  19cc30653e6da3a6d1ed90b62e93325b6dd3cd08 from
https://source.codeaurora.org/external/imx/linux-imx.git

There's no meaning to fake a wrong rate to recalc.
Instread, simply return 0 for this case.

Before:
    spll_pre_sel                          1            1    24000000          0 0
       spll_pre_div                       1            1    24000000          0 0
          spll                            2            2   531648000          0 0
             spll_pfd3                    0            0   979729408          0 0
             spll_pfd2                    0            0   979729408          0 0
             spll_pfd1                    0            0   979729408          0 0
             spll_pfd0                    1            1   416072347          0 0

After:
    spll_pre_sel                          1            1    24000000          0 0
       spll_pre_div                       1            1    24000000          0 0
          spll                            2            2   531648000          0 0
             spll_pfd3                    0            0           0          0 0
             spll_pfd2                    0            0           0          0 0
             spll_pfd1                    0            0           0          0 0
             spll_pfd0                    1            1   416072347          0 0

Fixes: 78ef764871d6 ("MLK-13441-5 ARM: imx: add new clk types")
Cc: Anson Huang <Anson.Huang@nxp.com>
Reviewed-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Dong Aisheng <aisheng.dong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clk/imx/clk-pfdv2.c |   14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/clk/imx/clk-pfdv2.c b/drivers/clk/imx/clk-pfdv2.c
index 141ec611..fb44997 100644
--- a/drivers/clk/imx/clk-pfdv2.c
+++ b/drivers/clk/imx/clk-pfdv2.c
@@ -62,12 +62,11 @@ static unsigned long clk_pfd_recalc_rate(struct clk_hw *hw,
 	u64 tmp = parent_rate;
 	u8 frac = (readl_relaxed(pfd->reg) >> (pfd->idx * 8)) & 0x3f;
 
-	/*
-	 * The reset value of pfd field is zero, so add one to avoid div
-	 * by zero, optimize this late.
-	 */
-	if (!frac)
-		frac += 1;
+	if (!frac) {
+		pr_debug("clk_pfdv2: %s invalid pfd frac value 0\n",
+			 clk_hw_get_name(hw));
+		return 0;
+	}
 
 	tmp *= 18;
 	do_div(tmp, frac);
@@ -103,6 +102,9 @@ static int clk_pfd_set_rate(struct clk_hw *hw, unsigned long rate,
 	u32 val;
 	u8 frac;
 
+	if (!rate)
+		return -EINVAL;
+
 	/* PFD can NOT change rate without gating */
 	WARN_ON(!(readl_relaxed(pfd->reg) &
 		(1 << ((pfd->idx + 1) * 8 - 1))));
-- 
1.7.9.5

